lib/ansible/plugins/loader.py

```diff
@@ -463,11 +463,14 @@ def all(self, *args, **kwargs):
         if path_only and class_only:
             raise AnsibleError('Do not set both path_only and class_only when calling PluginLoader.all()')
 
-        all_matches = []
         found_in_cache = True
-
-        for i in self._get_paths():
-            all_matches.extend(glob.glob(os.path.join(i, "*.py")))
+        all_matches = [
+            os.path.join(dirpath, name)
+            for dirpath in self._get_paths()
+            if os.path.isdir(dirpath)
+            for name in os.listdir(dirpath)
+            if name[0] != '.' and name.endswith('.py')
+        ]
```

Comment:
This breaks configured paths that have globbing characters

Reply:
@pilou- @dw Hi, Is this something that you could please look into?

Reply:
Wonder if we need to add any tests for this?

Reply:
Is there an example or doc someplace where that actually happens? If glob.glob() is really needed, perhaps a better approach would be to avoid running this function so much, rather than attempt to replace glob.glob()

Reply:
You can specify a path glob to configure search paths for plugins.

Like the following contrived example:

```
ANSIBLE_CALLBACK_PLUGINS=`pwd`/*/ ansible localhost -m ping
```

Because we use `glob.glob` currently this allows for supplying plugin paths with glob chars, which people do use.

Reply:
Now it makes sense.. in this case it looks like it needs to happen in `_get_paths()` instead of or in addition to `all()`: ideally the globs would happen only once and remain cached while the lookup path remains unchanged, but that's a bunch more work than the original PR.

Happy to abandon this for now, unless someone else is volunteering :)

Reply:
I think the only change to glob in `_get_paths()` would be relatively simple:

```diff
diff --git a/lib/ansible/plugins/loader.py b/lib/ansible/plugins/loader.py
index d6cec2bf66..8611016adf 100644
--- a/lib/ansible/plugins/loader.py
+++ b/lib/ansible/plugins/loader.py
@@ -161,15 +161,15 @@ class PluginLoader:
 
         # look in any configured plugin paths, allow one level deep for subcategories
         if self.config is not None:
-            for path in self.config:
-                path = os.path.realpath(os.path.expanduser(path))
-                if subdirs:
-                    contents = glob.glob("%s/*" % path) + glob.glob("%s/*/*" % path)
-                    for c in contents:
-                        if os.path.isdir(c) and c not in ret:
-                            ret.append(c)
-                if path not in ret:
-                    ret.append(path)
+            for item in self.config:
+                for path in glob.glob(os.path.realpath(os.path.expanduser(item))):
+                    if subdirs:
+                        contents = glob.glob("%s/*" % path) + glob.glob("%s/*/*" % path)
+                        for c in contents:
+                            if os.path.isdir(c) and c not in ret:
+                                ret.append(c)
+                    if path not in ret:
+                        ret.append(path)
 
         # look for any plugins installed in the package subtree
         # Note package path always gets added last so that every other type of
```

Reply:
A few stats with this change compared to devel and this PR (100 hosts and 1 debug using jinja2):

```
==> devel.txt <==
         765719 function calls (758704 primitive calls) in 4.990 seconds

==> 45568.txt <==
         651530 function calls (644525 primitive calls) in 4.862 seconds

==> glob_get_paths.txt <==
         651709 function calls (644704 primitive calls) in 4.881 seconds
```

```
devel.txt
11:     3648    0.104    0.000    1.195    0.000 loader.py:423(all)
174:        2    0.000    0.000    0.030    0.015 loader.py:546(all)

45568.txt
16:     3648    0.069    0.000    0.734    0.000 loader.py:423(all)
218:        2    0.000    0.000    0.016    0.008 loader.py:549(all)

glob_get_paths.txt
17:     3648    0.051    0.000    0.724    0.000 loader.py:423(all)
211:        2    0.000    0.000    0.017    0.009 loader.py:549(all)
```

```
devel.txt
18:     2850    0.015    0.000    0.724    0.000 glob.py:9(glob)

45568.txt
451:       38    0.000    0.000    0.002    0.000 glob.py:9(glob)

glob_get_paths.txt
426:       38    0.000    0.000    0.002    0.000 glob.py:9(glob)
```